slasher/service/data_update.go

```diff
@@ -37,3 +46,134 @@ func (s *Service) finalisedChangeUpdater() error {
+			QueryFilter: &ethpb.ListCommitteesRequest_Epoch{
+				Epoch: i,
+			},
+		})
+		if err != nil || bcs == nil {
+			log.Error(err)
+			continue
+		}
+		log.Infof("detecting slashable events on: %v attestations from epoch: %v", len(ats.Attestations), i)
+		for _, attestation := range ats.Attestations {
+			scs, ok := bcs.Committees[attestation.Data.Slot]
+			if !ok {
+				var keys []uint64
+				for k := range bcs.Committees {
+					keys = append(keys, k)
+				}
+				log.Errorf("committees doesnt contain the attestation slot: %d, actual first slot: %v", attestation.Data.Slot, keys)
+				continue
+			}
+			if attestation.Data.CommitteeIndex > uint64(len(scs.Committees)) {
+				log.Errorf("committee index is out of range in slot wanted: %v, actual", attestation.Data.CommitteeIndex, len(scs.Committees))
+				continue
+			}
+			sc := scs.Committees[attestation.Data.CommitteeIndex]
+			c := sc.ValidatorIndices
+			ia, err := ConvertToIndexed(s.context, attestation, c)
+			if err != nil {
+				log.Error(err)
+				continue
+			}
+			sar, err := s.slasher.IsSlashableAttestation(s.context, ia)
+			if err != nil {
+				log.Error(err)
+				continue
+			}
+			s.slasherDb.SaveAttesterSlashings(db.Active, sar.AttesterSlashing)
+			if len(sar.AttesterSlashing) > 0 {
+				log.Infof("slashing response: %v", sar.AttesterSlashing)
+			}
+		}
+		s.slasherDb.SetLatestEpochDetected(i)
+		ch, err = s.beaconClient.GetChainHead(s.context, &ptypes.Empty{})
+		if err != nil {
+			log.Error(err)
+			s.Stop()
+		}
+	}
+	errorWg.Wait()
+	close(errOut)
+	for err := range errOut {
+		log.Error(errors.Wrap(err, "error while writing to db in background"))
+	}
+	return nil
+}
+
+// ConvertToIndexed converts attestation to (almost) indexed-verifiable form.
+//
+// Note about spec pseudocode definition. The state was used by get_attesting_indices to determine
+// the attestation committee. Now that we provide this as an argument, we no longer need to provide
+// a state.
+//
+// Spec pseudocode definition:
+//   def get_indexed_attestation(state: BeaconState, attestation: Attestation) -> IndexedAttestation:
+//    """
+//    Return the indexed attestation corresponding to ``attestation``.
+//    """
+//    attesting_indices = get_attesting_indices(state, attestation.data, attestation.aggregation_bits)
+//    custody_bit_1_indices = get_attesting_indices(state, attestation.data, attestation.custody_bits)
+//    assert custody_bit_1_indices.issubset(attesting_indices)
+//    custody_bit_0_indices = attesting_indices.difference(custody_bit_1_indices)
+//
+//    return IndexedAttestation(
+//        custody_bit_0_indices=sorted(custody_bit_0_indices),
+//        custody_bit_1_indices=sorted(custody_bit_1_indices),
+//        data=attestation.data,
+//        signature=attestation.signature,
+//    )
+func ConvertToIndexed(ctx context.Context, attestation *ethpb.Attestation, committee []uint64) (*ethpb.IndexedAttestation, error) {
```

Comment:
This is duplicated code between slasher and the beacon node. If this is the case, we could consider moving to a `shared/attestationutil` package in the Prysm top-level directory